homeassistant/components/recorder/table_managers/statistics_meta.py

```diff
@@ -0,0 +1,317 @@
+"""Support managing StatesMeta."""
+from __future__ import annotations
+
+from collections.abc import Callable
+import logging
+import threading
+from typing import TYPE_CHECKING, Literal, cast
+
+from lru import LRU  # pylint: disable=no-name-in-module
+from sqlalchemy import lambda_stmt, select
+from sqlalchemy.orm.session import Session
+from sqlalchemy.sql.expression import true
+from sqlalchemy.sql.lambdas import StatementLambdaElement
+
+from ..db_schema import StatisticsMeta
+from ..models import StatisticMetaData
+from ..util import execute_stmt_lambda_element
+
+if TYPE_CHECKING:
+    from ..core import Recorder
+
+CACHE_SIZE = 8192
+
+_LOGGER = logging.getLogger(__name__)
+
+QUERY_STATISTIC_META = (
+    StatisticsMeta.id,
+    StatisticsMeta.statistic_id,
+    StatisticsMeta.source,
+    StatisticsMeta.unit_of_measurement,
+    StatisticsMeta.has_mean,
+    StatisticsMeta.has_sum,
+    StatisticsMeta.name,
+)
+
+
+def _generate_get_metadata_stmt(
+    statistic_ids: list[str] | None = None,
+    statistic_type: Literal["mean"] | Literal["sum"] | None = None,
+    statistic_source: str | None = None,
+) -> StatementLambdaElement:
+    """Generate a statement to fetch metadata."""
+    stmt = lambda_stmt(lambda: select(*QUERY_STATISTIC_META))
+    if statistic_ids:
+        stmt += lambda q: q.where(
+            # https://github.com/python/mypy/issues/2608
+            StatisticsMeta.statistic_id.in_(statistic_ids)  # type:ignore[arg-type]
+        )
+    if statistic_source is not None:
+        stmt += lambda q: q.where(StatisticsMeta.source == statistic_source)
+    if statistic_type == "mean":
+        stmt += lambda q: q.where(StatisticsMeta.has_mean == true())
+    elif statistic_type == "sum":
+        stmt += lambda q: q.where(StatisticsMeta.has_sum == true())
+    return stmt
+
+
+def _generate_filter(
+    statistic_type: Literal["mean"] | Literal["sum"] | None = None,
+    statistic_source: str | None = None,
+) -> Callable[[tuple[int, StatisticMetaData]], bool]:
+    """Generate a filter function for metadata."""
+    if not statistic_type and not statistic_source:
+        return lambda _: True
+
+    def _filter(id_meta: tuple[int, StatisticMetaData]) -> bool:
+        meta = id_meta[1]
+        if statistic_source is not None and meta["source"] != statistic_source:
+            return False
+        if statistic_type == "mean" and not meta["has_mean"]:
+            return False
+        if statistic_type == "sum" and not meta["has_sum"]:
+            return False
```

Comment:
While it works fine, this is never used.

I didn't realize the arguments are mutually exclusive.

I'll change it to raise a ValueError instead so nobody does that in the future